fleetspeak/src/server/https/streaming_message_server.go

```diff
@@ -0,0 +1,391 @@
+	if err = proto.Unmarshal(buf, &wcd); err != nil {
+		return nil, makeError(fmt.Sprintf("error parsing body: %v", err), http.StatusBadRequest)
+	}
+
+	info, toSend, err := s.fs.InitializeConnection(ctx, addr, key, &wcd, true)
+	if err == comms.ErrNotAuthorized {
+		return nil, makeError("not authorized", http.StatusServiceUnavailable)
+	}
+	if err != nil {
+		return nil, makeError(fmt.Sprintf("error processing contact: %v", err), http.StatusInternalServerError)
+	}
+	pi.CacheHit = info.Client.Cached
+
+	out := proto.NewBuffer(make([]byte, 0, 1024))
+	// EncodeMessage prepends the size as a Varint.
+	if err := out.EncodeMessage(toSend); err != nil {
+		return nil, makeError(fmt.Sprintf("error preparing messages: %v", err), http.StatusInternalServerError)
+	}
+	st = time.Now()
+	sz, err := res.Write(out.Bytes())
+	if err != nil {
+		return nil, makeError(fmt.Sprintf("error writing body: %v", err), http.StatusInternalServerError)
+	}
+	res.Flush()
+
+	pi.WriteTime = time.Since(st)
+	pi.End = time.Now()
+	pi.WriteBytes = sz
+	pi.Status = http.StatusOK
+	return info, nil
+}
+
+type streamManager struct {
+	ctx context.Context
+	s   streamingMessageServer
+
+	info *comms.ConnectionInfo
+	res  fullResponseWriter
+	body *bufio.Reader
+
+	// The read- and writeLoop will wait for these. Separate because readloop
+	// needs to finish before writeLoop.
+	reading sync.WaitGroup
+	writing sync.WaitGroup
+
+	out chan *fspb.ContactData
+
+	cancel func() // Shuts down the stream when called.
+}
+
+func (m *streamManager) readLoop() {
+	defer m.reading.Done()
+	defer m.cancel()
+
+	cnt := uint64(0)
+
+	for {
+		pi, err := m.readOne()
+		if err != nil {
+			// If the context has been canceled, it is probably a 'normal' termination
+			// - disconnect, max connection durating, etc. But if it is still active,
+			// we are going to tear down everything because of an unexpected read
+			// error and should log/record why.
+			if m.ctx.Err() == nil && pi != nil {
+				m.s.fs.StatsCollector().ClientPoll(*pi)
+				log.Errorf("Streaming Connection to %v terminated with error: %v", m.info.Client.ID, err)
+			}
+			return
+		}
+		m.s.fs.StatsCollector().ClientPoll(*pi)
+		cnt += 1
+		m.out <- &fspb.ContactData{AckIndex: cnt}
+	}
+}
+
+func (m *streamManager) readOne() (*stats.PollInfo, error) {
+	size, err := binary.ReadUvarint(m.body)
+	if err != nil {
+		return nil, err
+	}
+	if size > MaxContactSize {
+		return nil, fmt.Errorf("streaming contact size too large: got %d, expected less than %d", size, MaxContactSize)
+	}
+
+	pi := stats.PollInfo{
+		CTX:      m.ctx,
+		ID:       m.info.Client.ID,
+		Start:    db.Now(),
+		Status:   http.StatusTeapot,
+		CacheHit: true,
+	}
+	defer func() {
+		if pi.Status == http.StatusTeapot {
+			log.Errorf("Forgot to set status.")
+		}
+		pi.End = db.Now()
+	}()
+	buf := make([]byte, size)
+	if _, err := io.ReadFull(m.body, buf); err != nil {
+		return &pi, fmt.Errorf("Error reading streamed data: %v", err)
```

Comment:
nit

The linter will complain about starting the error message in uppercase

Reply:
Done